File: feature/movie_GenAI/generate_images.py
import os
import openai
from datetime import datetime, timedelta
import supabase
from dotenv import load_dotenv
import base64
import requests
from PIL import Image
import io

# 環境変数の読み込み
load_dotenv()

# Supabaseの設定
supabase_url = os.getenv("NEXT_PUBLIC_SUPABASE_URL")
supabase_key = os.getenv("NEXT_PUBLIC_SUPABASE_ANON_KEY")
supabase_client = supabase.create_client(supabase_url, supabase_key)

# 画像生成のAPIエンドポイント
url = "https://api.openai.com/v1/images/generations"

# 認証用のヘッダー
headers = {
    "Authorization": f"Bearer {os.getenv('OPENAI_API_KEY')}",
    "Content-Type": "application/json"
}

# 過去1ヶ月分のイベントを取得
# 現状uid固定
def get_monthly_events(uid,days):
    end_date = datetime.now()
    start_date = end_date - timedelta(days)
    
    response = supabase_client.table('Action').select(
        'action_name, happiness_change, Calendar(timestamp), User!Action_uid_fkey(gender)'
    ).eq('uid', uid).gte('Calendar.timestamp', start_date.isoformat()).lte('Calendar.timestamp', end_date.isoformat()).execute()
    print(response.data)
    return response.data

# イベントから画像生成用のプロンプトを作成
def generate_prompt(event):
    prompt = "あなたはカップルのスライドショーの一部となる画像を生成するアシスタントです。"
    happiness = "嬉しい" if event['happiness_change'] > 0 else "悲しい"
    happiness_change = event['happiness_change']
    gender = "男性" if event['User']['gender'] == 'male' else "女性"
    prompt += f"性別：{gender}\n"
    prompt += f"{happiness}度合い： {happiness_change}\n"
    prompt += f"されたこと：{event['action_name']}\n"
    prompt += "以上の出来事を表現した画像を生成してください。アニメ調でお願いします。"
    return prompt

def generate_image_with_gpt(prompt):
    # GPT-4 Vision APIを使用して画像を生成
    try:
        # ペイロードを定義する
        payload = {
            "model": "gpt-image-1",
            "prompt": prompt,
            "n": 1,  # 生成する画像の枚数
            "size": "1024x1024",  # 画像の解像度
        }

        # リクエストを送信
        response = requests.post(url, headers=headers, json=payload)

        # レスポンスを確認
        if response.status_code == 200:
            data = response.json()
            # 画像をダウンロード
            image_b64 = data['data'][0]['b64_json']
            image_bytes = base64.b64decode(image_b64)
            image = Image.open(io.BytesIO(image_bytes))
            return image
    except Exception as e:
        print(f"Error generating image: {e}")
        return None

# イベントから画像を生成
def generate_images(uid, days):
    print(f"Generating images for user {uid} with days={days}")
    
    # 出力ディレクトリの作成
    output_dir = "feature/movie_GenAI/generated_images"
    os.makedirs(output_dir, exist_ok=True)
    
    # イベントの取得
    events = get_monthly_events(uid, days)
    
    # 各イベントに対して画像を生成
    for i, event in enumerate(events):
        prompt = generate_prompt(event)
        image = generate_image_with_gpt(prompt)
        
        if image:
            # 画像の保存
            action_name = event['action_name'].replace(' ', '').replace('/', '')
            happiness_change = event['happiness_change']
            timestamp = event['Calendar'][0]['timestamp']
            date_str = datetime.fromisoformat(timestamp).strftime("%Y%m%d_%H%M")
            image_path = os.path.join(output_dir, f"{action_name}_{happiness_change}_{date_str}_{i}.png")
            image.save(image_path)
            
            print(f"Generated image for event: {event['action_name']}")
        else:
            print(f"Failed to generate image for event: {event['action_name']}")

if __name__ == "__main__":
    # テスト用
    import sys
    if len(sys.argv) > 2:
        uid = sys.argv[1]
        days = int(sys.argv[2])
        generate_images(uid, days)
    else:
        print("Usage: python generate_images.py <uid> <days>")

File: feature/movie_GenAI/generateImages.js
const { createClient } = require('@supabase/supabase-js');
const { OpenAI } = require('openai');
const sharp = require('sharp');
const fs = require('fs').promises;
const path = require('path');
require('dotenv').config();

// OpenAIの設定
const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY
});

// Supabaseの設定
const supabase = createClient(
  process.env.SUPABASE_URL,
  process.env.SUPABASE_KEY
);

async function getMonthlyEvents() {
  const endDate = new Date();
  const startDate = new Date(endDate);
  startDate.setDate(startDate.getDate() - 30);

  const { data, error } = await supabase
    .from('Action')
    .select(`
      action_name,
      happiness_change,
      Calendar (
        timestamp
      )
    `)
    .gte('Calendar.timestamp', startDate.toISOString())
    .lte('Calendar.timestamp', endDate.toISOString());

  if (error) {
    console.error('Error fetching events:', error);
    return [];
  }

  return data;
}

function generatePrompt(event) {
  const happiness = event.happiness_change > 0 ? "幸せな" : "悲しい";
  return `${happiness} ${event.action_name}の様子を表現した画像を生成してください。高品質で写実的な画像を生成してください。`;
}

async function generateImageWithGPT(prompt) {
  try {
    const response = await openai.images.generate({
      model: "gpt-image-1",
      prompt: prompt,
      size: "1024x1024",
    //   quality: "standard",
      n: 1,
    });

    const imageUrl = response.data[0].url;
    const imageResponse = await fetch(imageUrl);
    const imageBuffer = await imageResponse.arrayBuffer();

    return Buffer.from(imageBuffer);
  } catch (error) {
    console.error('Error generating image:', error);
    return null;
  }
}

async function generateImages() {
  // 出力ディレクトリの作成
  const outputDir = path.join(__dirname, 'generated_images');
  await fs.mkdir(outputDir, { recursive: true });

  // イベントの取得
  const events = await getMonthlyEvents();

  // 各イベントに対して画像を生成
  for (let i = 0; i < events.length; i++) {
    const event = events[i];
    const prompt = generatePrompt(event);
    const imageBuffer = await generateImageWithGPT(prompt);

    if (imageBuffer) {
      // 画像の保存
      const timestamp = event.Calendar[0].timestamp;
      const dateStr = new Date(timestamp).toISOString().replace(/[:.]/g, '').slice(0, 15);
      const imagePath = path.join(outputDir, `event_${dateStr}_${i}.png`);

      // 画像をリサイズして保存
      await sharp(imageBuffer)
        .resize(1920, 1080, { fit: 'inside' })
        .toFile(imagePath);

      console.log(`Generated image for event: ${event.action_name}`);
    } else {
      console.log(`Failed to generate image for event: ${event.action_name}`);
    }
  }
}

module.exports = { generateImages }; 